oauth/src/lib.rs

```diff
@@ -0,0 +1,227 @@
+//! Provides a Spotify access token using the OAuth authorization code flow
+//! with PKCE.
+//!
+//! Assuming sufficient scopes, the returned access token may be used with Spotify's
+//! Web API, and/or to establish a new Session with [`librespot_core`].
+//!
+//! The authorization code flow is an interactive process which requires a web browser
+//! to complete. The resulting code must then be provided back from the browser to this
+//! library for exchange into an access token. Providing the code can be automatic via
+//! a spawned http server (mimicking Spotify's client), or manually via stdin. The latter
+//! is appropriate for headless systems.
+
+use log::{error, info, trace};
+use oauth2::reqwest::http_client;
+use oauth2::{
+    basic::BasicClient, AuthUrl, AuthorizationCode, ClientId, CsrfToken, PkceCodeChallenge,
+    RedirectUrl, Scope, TokenResponse, TokenUrl,
+};
+use std::io;
+use std::time::{Duration, Instant};
+use std::{
+    io::{BufRead, BufReader, Write},
+    net::{IpAddr, Ipv4Addr, SocketAddr, TcpListener},
+    sync::mpsc,
+};
+use thiserror::Error;
+use url::Url;
+
+#[derive(Debug, Error)]
+pub enum OAuthError {
```

Comment:
Would you also provide an `impl From<SpircError> for Error`?

Reply:
Sorry, not sure I follow. How/why would this be used here?

Reply:
Sorry, I meant `impl From<OAuthError> for core::Error` like so:

https://github.com/librespot-org/librespot/blob/f6473319f6309b2702a20ff5371763c48e129676/connect/src/spirc.rs#L48

Reply:
The only thing not ideal with that, is having to pull in librespot `core` to do it..

Reply:
Possibly the other way around: put the implementation in core::error so that pulls in oauth.